### ISSUE TITLE
A first approach to impedance control

### ISSUE THREAD

**Issue by PeterBowman** [roboticslab-uc3m/kinematics-dynamics#163]
TEO has recently learned to behave like a spring (boing, boing), or at least with one of its upper limbs rather than using the whole body. So far, it integrates Hooke's law to obtain the resulting external force to be applied on the TCP while a human operator drives one of its hands off an initial pose during gravity compensation control. I fact, all it does is to multiply the displacement between such pose and its current position by a stiffness. Then, the contribution of the resulting vector of forces is used to perform inverse dynamics and, adding to the computed torques needed to keep the arm in place (gravity compensation), another contribution helps to make the TCP travel back with a spring-like effect.

Much to TEO's disgrace, a suitable value of stiffness has been estimated through trial and error to range between 100-150. Anything higher than that requires extra care and good reflexes.

At the time of writing, only linear forces are taken into account in the current implementation: https://github.com/roboticslab-uc3m/kinematics-dynamics/compare/roboticslab-uc3m:6e78a53...roboticslab-uc3m:b8ed4ca. Torques (in cartesian space) could be easily introduced thanks to [`ICartesianSolver::poseDiff`](https://github.com/roboticslab-uc3m/kinematics-dynamics/blob/6e78a535538298730670d53cd8206d504bc01a46/libraries/YarpPlugins/ICartesianSolver.h#L115), but keep in mind that [torsional stiffness](https://en.wikipedia.org/wiki/Stiffness#Rotational_stiffness) should be used in such cases instead.

We observed that it's hard to get back the initial pose after releasing the arm. Also, if higher displacements and stiffness values are applied, it behaves like an oscillator, hence some sort of damping effect could be of use.

In addition to the previous ideas, the torques required to undo the displacement (in cartesian space, considering a sole external force applied on the TCP) might be obtained through the same Hooke's law, but now assuming it behaves like several [torsion springs](https://en.wikipedia.org/wiki/Torsion_spring) instead. That is, measure the angles (one per joint) between initial and current pose, then use the resulting torque in the outcome of inverse dynamics.

**Comment by PeterBowman** [comment]
TODOs:

* investigate iCUB/YARP usage of the position/velocity impedance mode from `yarp::dev::IControlMode` ([ref](https://github.com/robotology/yarp/blob/62d24814a32a7fb4f30c8c0eeaf3247111178853/src/libYARP_dev/include/yarp/dev/IControlMode.h#L134-L135), deprecated long ago)
  * perhaps superseded by [`yarp::dev::InteractionMode`](http://www.yarp.it/classyarp_1_1dev_1_1IInteractionMode.html)
* investigate [`yarp::dev::ImpedanceControl`](http://www.yarp.it/classyarp_1_1dev_1_1IImpedanceControl.html)
* what about friction? https://github.com/robotology-playground/torque-control-params-estimation
* how to combine gravity compensation and internal friction forces? http://roboticslab.uc3m.es/roboticslab/book/advances-intelligent-systems-and-computing-vol-418 (thanks, @jgvictores)

**Comment by PeterBowman** [comment]
Watch and limit velocities? See https://github.com/roboticslab-uc3m/kinematics-dynamics/issues/163.

**Comment by PeterBowman** [comment]
I'm going to transfer this issue from the kinematics-dynamics repo to yarp-devices. I believe this problem should be tackled in the joint space, not in the cartesian space.

First, the modifications to BasicCartesianControl proposed at https://github.com/roboticslab-uc3m/kinematics-dynamics/compare/6e78a53...b8ed4ca are a hackish way to demonstrate impedance control, and nothing more than that. It has been implemented on top of the gravity compensation command for simplicity, but it defeats the real purpose of gcmp. In fact, if I were to implement a spring-like behavior while maintaining position, I'd rather configure the manipulator in position control mode and enable the compliant behavior via `IInteractionMode`.

Secondly, imposing compliance on a per-joint manner (joint space) is quite equivalent to simulating a spring-like response on external forces applied on the TCP (cartesian space), which need to be translated back to joint space via inverse dynamics anyway. Quite, but not really, since our solution doesn't account for forces applied elsewhere, i.e. on intermediate joints ant links. Besides, we should account for changes in orientation as well (we didn't implement this on the testing branch and almost made the robot collide with itself when the arm started to oscillate).

In conclusion, although it might be of interest to revisit cartesian-space impedance control at some point, it is more reasonable to have a working joint-space implementation under the hood (also, that should be way easier).
